edx_event_bus_kafka/internal/producer.py

```diff
@@ -191,28 +234,42 @@ def send(
-
-        key_serializer, value_serializer = get_serializers(signal, event_key_field)
-        key_bytes = key_serializer(event_key, SerializationContext(full_topic, MessageField.KEY, headers))
-        value_bytes = value_serializer(event_data, SerializationContext(full_topic, MessageField.VALUE, headers))
-
-        self.producer.produce(
-            full_topic, key=key_bytes, value=value_bytes, headers=headers, on_delivery=on_event_deliver,
-        )
-
-        # Opportunistically ensure any pending callbacks from recent event-sends are triggered.
-        # This ensures that we're polling at least as often as we're producing, which is a
-        # reasonable balance. However, if events are infrequent, it doesn't ensure that
-        # callbacks happen in a timely fashion, and the last event emitted before shutdown
-        # would never get a delivery callback. That's why there's also a thread calling
-        # poll(0) on a regular interval (see `poll_indefinitely`).
-        self.producer.poll(0)
+        # keep track of the initial arguments for recreating the event in the logs if necessary later
+        context = ProducingContext(context_data={SIGNAL_DICT_KEY: signal, INITIAL_TOPIC_DICT_KEY: topic,
+                                                 EVENT_KEY_FIELD_DICT_KEY: event_key_field,
+                                                 EVENT_DATA_DICT_KEY: event_data})
+        try:
+            full_topic = get_full_topic(topic)
+            context.set_context_data(FULL_TOPIC_DICT_KEY, full_topic)
+
+            event_key = extract_event_key(event_data, event_key_field)
+            context.set_context_data(EVENT_KEY_DICT_KEY, event_key)
+            headers = {EVENT_TYPE_HEADER_KEY: signal.event_type.encode("utf-8")}
+            context.set_context_data(EVENT_TYPE_DICT_KEY, signal.event_type)
+
+            key_serializer, value_serializer = get_serializers(signal, event_key_field)
+            key_bytes = key_serializer(event_key, SerializationContext(full_topic, MessageField.KEY, headers))
+            value_bytes = value_serializer(event_data, SerializationContext(full_topic, MessageField.VALUE, headers))
+            self.producer.produce(
+                full_topic, key=key_bytes, value=value_bytes, headers=headers,
+                on_delivery=context.on_event_deliver,
+            )
+
+            # Opportunistically ensure any pending callbacks from recent event-sends are triggered.
+            # This ensures that we're polling at least as often as we're producing, which is a
+            # reasonable balance. However, if events are infrequent, it doesn't ensure that
+            # callbacks happen in a timely fashion, and the last event emitted before shutdown
+            # would never get a delivery callback. That's why there's also a thread calling
+            # poll(0) on a regular interval (see `poll_indefinitely`).
+            self.producer.poll(0)
+        except BaseException as e:
```

Comment:
I am going to make some `record_exception()` requests, that I'd like added to the ticket. Feel free to spin some or all of this to a new ticket. It's simple to discuss here because we are looking at all the right places.

In this case, we are in an exception handler, so we just need a simple call to `record_exception()`. See [example call](https://github.com/openedx/edx-enterprise/blob/0f21072f25f2a5f2fda7117e1c260fc148a657d5/enterprise/views.py#L708).

Reply:
[fixed]